src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
+
+	return strncat(str1, str2, len);
+}
+
+const char *internal_keys[] = { "HOME", "CFG", "RUNUSER", "PATH", "PRIVILEGED",
+				NULL };
+
+static int is_internal_macro(const char *key)
+{
+	char *macro;
+	char *macro_value;
+	int err = 0;
+	int i;
+
+	for (i = 0; internal_keys[i]; i++) {
+		if (!strncmp(key, internal_keys[i], strlen(internal_keys[i]))) {
+			err = -EEXIST;
+			goto exit;
+		}
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	macro_value = resolve_macro(key);
+	if (macro_value) {
+		err = -EEXIST;
+		free(macro_value);
+	}
+
+	free(macro);
+exit:
+	return err;
+}
+
+static char *process_key_value(char *container, char *key, int *err)
+{
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_SUB_DELIMS;
+	const char *value;
+
+	for (str = key; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
+		if (!token)
+			break;
+
+		if (is_internal_macro(token) == -EEXIST) {
+			char *macro;
+
+			if (asprintf(&macro, "${%s}", token) == -1)
+				errExit("asprintf");
+
+			container = append_to_string(container, macro);
+			free(macro);
+			continue;
+		}
+
+		/* Only the first iteration is the template */
+		value = str ? template_get_value(token, err) : token;
+		if (!value)
+			goto err;
+
+		container = append_to_string(container, value);
+		if (!container)
+			goto err;
+	}
+
+	return container;
+
+err:
+	if (container)
+		free(container);
+	else
+		*err = -EINVAL;
+
+	return NULL;
+}
+
+/*
+ * Allocates new string with all template keus replaced with the values.
+ * Returns NULL if there is a nonexistent key, allocation fails or if arg
+ * begins with $. If arg does not contain $ it is only duplicated.
+ */
+char *template_replace_keys(char *arg, int *err)
+{
+	char *new_string = NULL;
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_DELIM;
+
+	if (!arg || !*arg) {
+		*err = -ENOENT;
+		return NULL;
+	}
+
+	if (!strrchr(arg, '$')) {
+		fprintf(stderr, "No \"$\" in \"%s\" return copy\n", arg);
+		return strdup(arg);
+	}
+
+	/* Templates must not be given at the beginning of the argument */
+	if (*arg == '$') {
+		*err = -EINVAL;
+		fprintf(stderr, "Starts with \"$\" in \"%s\"\n", arg);
+		return NULL;
+	}
+
+	for (str = arg; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
+		if (!token)
+			break;
+
+		/*
+		 * Process template values starting from the second token as
+		 * templates cannot be used at the beginning of the arg
+		 * because only hardcoded macros should be as first.
+		 */
+		if (!str)
+			new_string = process_key_value(new_string, token, err);
+		else
+			new_string = append_to_string(new_string, token);
+
+		if (!new_string)
+			goto err;
+	}
+
+	fprintf(stderr, "Return \"%s\"\n", new_string ? new_string : "<null>");
+	return new_string;
+
+err:
+	if (new_string)
+		free(new_string);
+	else if (!err)
+		*err = -ENOMEM;
```

Comment:
eh? if (err == NULL) *err = please_crash;

Also, going through all this passing error code pointers back and forth -> looks like it is uninitialized value originating from depths of profile_read() function -> it would not make sense for it to be 'if( *err ) *err = something:' either.

Reply:
oops a mistake there. Should be `else if (!*err)` thanks for pointing it out.

Reply:
... and that: what if *err is already non-zero due to originating from uninitialized-variable-in-stack like I think it is?

IMO you would either need to ensure that each and every error code is zero initialized, or not rely on previous content at all. In this case there is one place where reason is possibly ENOMEM and you could just set it unconditionally there and omit this implied ENOMEM bit?